overlord/snapstate/snapstate.go

```diff
@@ -782,10 +790,33 @@ func doInstall(st *state.State, snapst *SnapState, snapsup SnapSetup, compsups [
 	installSet := state.NewTaskSet(tasks...)
 	installSet.MarkEdge(prereq, BeginEdge)
 	installSet.MarkEdge(setupAliases, BeforeHooksEdge)
-	installSet.MarkEdge(setupSecurity, BeforeMaybeRebootEdge)
-	installSet.MarkEdge(linkSnap, MaybeRebootEdge)
-	installSet.MarkEdge(autoConnect, MaybeRebootWaitEdge)
-	installSet.MarkEdge(setAutoAliases, AfterMaybeRebootWaitEdge)
+
+	// restart related edges
+	// TODO fix tests so BeforeMaybeRebootEdge and AfterMaybeRebootWaitEdge
+	// are not needed.
+	installSet.MarkEdge(setupSecurity, BeforeMaybeRebootEdge) // this edge is just for tests
+	var restartInLink, restartInSetupKmods bool
+	if setupKmodComponents == nil {
+		installSet.MarkEdge(linkSnap, MaybeRebootEdge)
+		installSet.MarkEdge(autoConnect, MaybeRebootWaitEdge)
+		restartInLink = true
+	} else {
+		logger.Noticef("reboot will happen after set-up of kernel-modules")
+		installSet.MarkEdge(setupKmodComponents, MaybeRebootEdge)
+		installSet.MarkEdge(afterSetupKmodComps, MaybeRebootWaitEdge)
+		restartInSetupKmods = true
+	}
+	// Let tasks know if they have to do something about restarts
+	linkSnap.Set("prepare-restart", restartInLink)
+	autoConnect.Set("finish-restart", restartInLink)
+	if setupKmodComponents != nil {
+		setupKmodComponents.Set("prepare-restart", restartInSetupKmods)
+	}
+	if afterSetupKmodComps != nil {
+		afterSetupKmodComps.Set("finish-restart", restartInSetupKmods)
+	}
```

Comment:
Should we not just set those up in the `else` above? Then we can save the `restartInSetupKmods` bool?